Get all possible outcomes from a probabilistic effect

GCrispino opened the pull request:
I am creating this (currently WIP) PR that will implement what was discussed in issue #49.
As I add new commits and potentially have some comments/doubts, I'll add some notes here.

GCrispino commented:
For the first commit I created a simple probabilistic domain to test the new feature. I decided to basically copy the domain in test_domain.pddl and modify the single effect in for pred2 to a probabilistic effect with 50% probability.

I changed its name from pred2 to probpred because otherwise I got an AssertionError when executing the following code from the `pddlgym/tests` directory to try to load the domain:

```python
import os
from pddlgym.core import PDDLEnv
dir_path = os.path.dirname('.') 
domain_file = os.path.join(dir_path, 'pddl', 'test_probabilistic_domain.pddl') 
problem_dir = os.path.join(dir_path, 'pddl', 'test_probabilistic_domain') 
  
env = PDDLEnv(domain_file, problem_dir) 
```

In case you want to take a look, the error I get is the following:

```python
---------------------------------------------------------------------------                                                              
AssertionError                            Traceback (most recent call last)          
<ipython-input-223-1bc64143fdb9> in <module>                       
      3 problem_dir = os.path.join(dir_path, 'pddl', 'test_probabilistic_domain')                                                                
      4                                                                                                                                          
----> 5 env = PDDLEnv(domain_file, problem_dir)                                                                                                  
                                                                                                                                                 
~/code/Python/pddlgym/pddlgym/core.py in __init__(self, domain_file, problem_dir, render, seed, raise_error_on_invalid_action, operators_as_actio
ns, dynamic_action_space)                                                                                                                        
    240                                                                                                                                          
    241         # Parse the PDDL files                                                                                                           
--> 242         self.domain, self.problems = self.load_pddl(domain_file, problem_dir,
    243             operators_as_actions=self.operators_as_actions)                                                                              
    244                                                                                                                                          
                                                                                                                                                 
~/code/Python/pddlgym/pddlgym/core.py in load_pddl(domain_file, problem_dir, operators_as_actions)                                               
    292         problems : [ PDDLProblemParser ]                                                                                                 
    293         """                                                                                                                              
--> 294         domain = PDDLDomainParser(domain_file,                                                                                           
    295             expect_action_preds=(not operators_as_actions),                                                                              
    296             operators_as_actions=operators_as_actions)
                                                                        
~/code/Python/pddlgym/pddlgym/parser.py in __init__(self, domain_fname, expect_action_preds, operators_as_actions)                               
    453                                                                                                                                          
    454         # Run parsing.                                                                                                                   
--> 455         self._parse_domain()                                                                                                             
    456 
    457         if operators_as_actions:

~/code/Python/pddlgym/pddlgym/parser.py in _parse_domain(self)
    484         self._parse_domain_derived_predicates()
    485         self._parse_constants()
--> 486         self._parse_domain_operators()
    487 
    488     def _parse_domain_types(self):

~/code/Python/pddlgym/pddlgym/parser.py in _parse_domain_operators(self)
    611                 params = [self.types["default"]("?"+k) for k in params]
    612             preconds = self._parse_into_literal(preconds.strip(), params + self.constants)
--> 613             effects = self._parse_into_literal(effects.strip(), params + self.constants,
    614                 is_effect=True)
    615             self.operators[op_name] = Operator(

~/code/Python/pddlgym/pddlgym/parser.py in _parse_into_literal(self, string, params, is_effect)
    105         if string.startswith("(and") and string[4] in (" ", "\n", "("):
    106             clauses = self._find_all_balanced_expressions(string[4:-1].strip())
--> 107             return LiteralConjunction([self._parse_into_literal(clause, params, 
    108                                        is_effect=is_effect) for clause in clauses])
    109         if string.startswith("(or") and string[3] in (" ", "\n", "("):

~/code/Python/pddlgym/pddlgym/parser.py in <listcomp>(.0)
    105         if string.startswith("(and") and string[4] in (" ", "\n", "("):
    106             clauses = self._find_all_balanced_expressions(string[4:-1].strip())
--> 107             return LiteralConjunction([self._parse_into_literal(clause, params, 
    108                                        is_effect=is_effect) for clause in clauses])
    109         if string.startswith("(or") and string[3] in (" ", "\n", "("):

~/code/Python/pddlgym/pddlgym/parser.py in _parse_into_literal(self, string, params, is_effect)
    166             for match in re.finditer("(\d*\.?\d+)", expr):
    167                 prob = float(match.group())
--> 168                 subexpr = self._find_balanced_expression(expr[match.end():].strip(), 0)
    169                 lit = self._parse_into_literal(subexpr, params, is_effect=is_effect)
    170                 lits.append(lit)

~/code/Python/pddlgym/pddlgym/parser.py in _find_balanced_expression(string, index)
    260         """Find balanced expression in string starting from given index.
    261         """
--> 262         assert string[index] == "("
    263         start_index = index
    264         balance = 1

AssertionError: 
```

It looks like some parenthesis-matching error, but I found it weird that only changing the predicate's name prevented it.

Nonetheless, this is secondary and the next step is to start writing tests and the functionality.

ronuchit commented:
Thanks for finding this issue. It looks like what was happening is that when you have a number in the predicate name for a probabilistic effect, e.g. pred2, the "2" was getting parsed as a probability. I've changed the parsing to require a decimal point to appear in the probability value: https://github.com/tomsilver/pddlgym/commit/aa26f831c11fa84604bb8855a47a31ef83697735

Hopefully this should allow you to use pred2, which I would prefer over probpred. Let me know if it doesn't work.

GCrispino commented:
> Thanks for finding this issue. It looks like what was happening is that when you have a number in the predicate name for a probabilistic effect, e.g. pred2, the "2" was getting parsed as a probability. I've changed the parsing to require a decimal point to appear in the probability value: [aa26f83](https://github.com/tomsilver/pddlgym/commit/aa26f831c11fa84604bb8855a47a31ef83697735)

Great!

> Hopefully this should allow you to use pred2, which I would prefer over probpred. Let me know if it doesn't work.

Me too. Already changed it back to pred2.

GCrispino commented on pddlgym/core.py:
```diff
@@ -78,6 +79,36 @@ def get_successor_state(state, action, domain, raise_error_on_invalid_action=Fal
     return state
 
+
+def get_successor_states(state, action, domain, raise_error_on_invalid_action=False,
```
Together with the method `get_all_possible_transitions(self, action)` that was suggested, I found it valuable to add this function that gets the successor states from a passed state and action, instead of getting the state from the current env object.
`get_all_possible_transitions` then uses this by passing its current state.

ronuchit replied:
It looks like this method is the same as get_successor_state except that get_all_transitions=True in the call to _apply_effects. It might be better to just add an argument to get_successor_state, called get_all_possible or something, that is False by default? And then just pass that argument through to _apply_effects. I think this might help keep the code cleaner because then we don't need this new get_successor_states method. What do you think?

GCrispino replied:
That's true, I agree. My only concern would be minor naming issues (`get_successor_states` means more that possibly more states can be return instead of just one). What would ou think about keeping get_successor_states but only to call the new get_sucessor_state function with the new flag marked as True?

ronuchit replied:
Yeah, that sounds good!

GCrispino replied:
Awesome!

GCrispino commented on pddlgym/core.py:
```diff
@@ -417,11 +495,30 @@ def sample_transition(self, action):
+        state = self._get_successor_state(self._state, action, self.domain,
+                                          inference_mode=self._inference_mode,
+                                          raise_error_on_invalid_action=self._raise_error_on_invalid_action)
+        return self._get_new_state_info(state)
```
I took the liberty to refactor here by using the newly added `_get_new_state_info` function to avoid code duplication.

GCrispino commented on pddlgym/core.py:
```diff
@@ -155,7 +186,8 @@ def _check_struct_for_strips(struct):
 
-def _apply_effects(state, lifted_effects, assignments):
+
+def _apply_effects(state, lifted_effects, assignments, get_all_transitions=False):
```
This function is messed up, it still needs some refactoring after these new changes.

ronuchit replied:
Cool, I'll take a look after you've finished the refactoring.

GCrispino replied:
Finished refactoring. Let me know if you have any concerns :)

ronuchit replied:
Looks good to me! Please make sure tests pass, both the current ones and the one(s) that you add. After you fix the get_successor_states() issue above, we should be ready to merge.

GCrispino replied:
Fine! I've already added one test for this new functionality, did you see it?

ronuchit replied:
That test looks good, added a minor comment for it.

GCrispino commented on pddlgym/core.py:
```diff
@@ -417,11 +495,30 @@ def sample_transition(self, action):
         """
         return get_successor_state(*args, **kwargs)
 
+    def _get_successor_states(self, *args, **kwargs):
```
Don't really know if this is necessary, but I followed the format in `_get_successor_state`.

ronuchit replied:
Looks great!

GCrispino commented:
Hello!
I just added the new functionality with the recent commits. It still needs refactoring in the `_apply_effects` function. I just got it working and added tests, but I'll refactor it to finish the PR changes.

I also added some comments in the code in case you want to take a look. Feel free to ask any questions!

Thanks!

GCrispino commented on pddlgym/core.py:
```diff
@@ -171,31 +203,80 @@ def _apply_effects(state, lifted_effects, assignments):
+    # - Refactor code after finished
+    states = []
+    if not get_all_transitions:
+        for lifted_effect in determinized_lifted_effects:
+            if lifted_effect == "NOCHANGE":
+                continue
+            effect = ground_literal(lifted_effect, assignments)
+            # Negative effect
+            if effect.is_anti:
+                literal = effect.inverted_anti
+                if literal in new_literals:
+                    new_literals.remove(literal)
+        for lifted_effect in determinized_lifted_effects:
+            if lifted_effect == "NOCHANGE":
+                continue
+            effect = ground_literal(lifted_effect, assignments)
+            if not effect.is_anti:
+                new_literals.add(effect)
```
Is there any reason for these two for loops exist instead of just one with the if statements combined?

ronuchit replied:
Good question! Yes, this was something we did specifically to ensure that negative effects were processed before positive effects. We should make sure to keep it that way!

ronuchit commented on pddlgym/tests/test_pddlenv.py:
```diff
@@ -122,7 +122,41 @@ def test_pddlenv_hierarchical_types():
+
+    obs, _ = env.reset()
+    action = env.action_space.all_ground_literals(obs).pop()
+    states = env.get_all_possible_transitions(action)
```
Minor: this isn't actually a list of states, right? It's a list of (state, reward, done, debug) tuples, if I understand the code correctly. So maybe change variable name to "transitions" or something, so that we don't confuse others looking at this example in the future :)

GCrispino commented on pddlgym/core.py:
```diff
@@ -71,6 +73,7 @@ def get_successor_state(state, action, domain, raise_error_on_invalid_action=Fal
             state,
             effects,
             assignment,
+            get_all_transitions
```
What do you think about changing this variable name from "state" to something else? I thought about changing it to "states" but this is not entirely accurate since this might return a single state if the flag is False.
Other possible candidate that I thought of is "result", but this might be too general.

ronuchit replied:
Let's leave it as is, so that we don't break dependency code in case they are using keyword arguments when calling this function.

GCrispino commented:
Hello! Just added the new changes as we discussed. Just added a minor comment about variable naming, but other than that, I think its ready :)

ronuchit reviewed (approved):
I think this is ready to merge! Any final thoughts, @GCrispino? Also, @tomsilver please do a final review.

GCrispino commented:
> I think this is ready to merge! Any final thoughts, @GCrispino? Also, @tomsilver please do a final review.

I just noticed some that there are some unused variables, but since this is not really related to this PR I could open another one. So I think this is it :)

tomsilver reviewed (approved):
LGTM, thanks so much for the contribution!

GCrispino commented:
Great, thanks!

GCrispino commented:
Hi @ronuchit and @tomsilver !
Thinking better about this PR's functionality, would you think that instead of returning a set of successor states in get_successor_states, returning a list or set containing tuples that holds the new state and also the probability of reaching this state from the given state and action would be a better idea?

I ask this because this is important information and I don't think that there's currently an "easy" way to obtain this probabilities (maybe getting the resulting effects and then getting the probabilities).

One drawback of following this direction is that the effects' information is retrieved and used in the _apply_effects function. So, if the probabilities would be returned, maybe this would involve changing it so it returns a (states, probabilities) tuple instead of just a state.

Any thoughts?

ronuchit commented:
Hi @GCrispino,

Compared to before, I am in town now, so can help much more quickly with your suggestions :)

I've made the following PR: https://github.com/tomsilver/pddlgym/pull/51. Could you check if it's what you're looking for? (See my updates to your test function for usage examples.) If so, I can merge it into master.

Please feel free to make comments directly on the new PR.

Rohan

GCrispino commented:
Ok! Will answer you there :)
